### PR TITLE
[FLINK-8174] [mesos] Mesos RM unable to accept offers for unreserved resources

### PR DESCRIPTION
**Pull request by EronWright**
## What is the purpose of the change
This PR improves the Mesos RM to work correctly with both reserved and unreserved resources.  Closes FLINK-8174.

## Brief change log
- Introduce `MesosResourceAllocation` to assign offered resources to task resources, preserving reservation metadata.
- Introduce `Offer` to replace Fenzo's `VMLeaseObject` with proper accounting across reserved and unreserved resources.

## Verifying this change

This change added tests and can be verified as follows:
- Unit tests:
  - `MesosResourceAllocationTest`
  - `OfferTest`
- Manual tests:
  - Launch with unreserved resources
  - Launch with a static reservation
  - Launch with a dynamic reservation
  - Launch with a mix of reserved and unreserved resources
  - Launch numerous tasks from a single offer

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by EronWright** [comment]
Log output showing a 'mixed' scenario, involving two tasks, combining resources across the reserved and unreserved resources in a single offer.
```
Assigning resources: [cpus(myrole):2.0; mem(myrole):4096.0; ports(myrole):[1025-2180]; ports(*):[2182-3887,3889-5049,5052-8079,8082-8180,8182-32000]; disk(*):28829.0; cpus(*):2.0; mem(*):10766.0]   

Processing task taskmanager-00002                                                                                         
Allocating 1.5 cpus                                                                                                       
Taking 1.5 from cpus(myrole):2.0                                             
Allocated: [cpus(myrole):1.5], unsatisfied: 0.0                              
Allocating 4096.0 mem                                                                                                     
Taking 4096.0 from mem(myrole):4096.0                                        
Allocated: [mem(myrole):4096.0], unsatisfied: 0.0                            
Allocating 2 ports                                                                                                        
Taking [1025-1026] from ports(myrole):[1025-2180]                            
Allocated: [ports(myrole):[1025-1026]], unsatisfied: 0                       

Processing task taskmanager-00001                                                                                         
Allocating 1.5 cpus                                                                                                       
Taking 0.5 from cpus(myrole):0.5                                             
Taking 1.0 from cpus(*):2.0                                                                                               
Allocated: [cpus(myrole):0.5; cpus(*):1.0], unsatisfied: 0.0                 
Allocating 4096.0 mem                                                                                                     
Taking 4096.0 from mem(*):10766.0                                                                                         
Allocated: [mem(*):4096.0], unsatisfied: 0.0                                                                              
Allocating 2 ports                                                                                                        
Taking [1027-1028] from ports(myrole):[1027-2180]                            
Allocated: [ports(myrole):[1027-1028]], unsatisfied: 0                       

Remaining resources: [ports(myrole):[1029-2180]; ports(*):[2182-3887,3889-5049,5052-8079,8082-8180,8182-32000]; disk(*):28829.0; cpus(*):1.0; mem(*):6670.0]
```

**Comment by tillrohrmann** [comment]
Thanks for your fast contribution @EronWright. I will take a closer look at the fix but from a first glance it seems that the `LaunchCoordinatorTest` fails. Could you fix this test case while I look at the code? Thanks a lot.

File: flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala [hunk 1]
```diff
@@ -148,14 +149,17 @@ class LaunchCoordinator(
 
     case Event(offers: ResourceOffers, data: GatherData) =>
       val leases = offers.offers().asScala.map(
-        new VMLeaseObject(_).asInstanceOf[VirtualMachineLease])
+        new Offer(_).asInstanceOf[VirtualMachineLease])
```
[the remark this answers is not in the document]
Comment (EronWright) [reply flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala]: If I recall correctly, The collection must be of a certain type.

File: flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala [hunk 2]
```diff
@@ -148,14 +149,17 @@ class LaunchCoordinator(
...
       if(LOG.isInfoEnabled) {
         val (cpus, mem) = leases.foldLeft((0.0,0.0)) {
           (z,o) => (z._1 + o.cpuCores(), z._2 + o.memoryMB())
         }
         LOG.info(s"Received offer(s) of $mem MB, $cpus cpus:")
         for(l <- leases) {
-          LOG.info(s"  ${l.getId} from ${l.hostname()} of ${l.memoryMB()} MB, ${l.cpuCores()} cpus")
+          val reservations = l.asInstanceOf[Offer].getResources.asScala.map(_.getRole).toSet
```
[the remark this answers is not in the document]
Comment (EronWright) [reply flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala]: (see above)

File: flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java [hunk 3]
```diff
@@ -217,15 +222,13 @@ public String toString() {
...
-		for (int i = 0; i < TM_PORT_KEYS.length; i++) {
-			int port = assignment.getAssignedPorts().get(i);
-			String key = TM_PORT_KEYS[i];
-			taskInfo.addResources(ranges("ports", mesosConfiguration.frameworkInfo().getRole(), range(port, port)));
-			dynamicProperties.setInteger(key, port);
+		// take needed ports for the TM
+		List<Protos.Resource> portResources = allocation.takeRanges("ports", TM_PORT_KEYS.length, roles);
+		taskInfo.addAllResources(portResources);
+		Iterator<String> portsToAssign = Iterators.forArray(TM_PORT_KEYS);
+		rangeValues(portResources).forEach(port -> dynamicProperties.setLong(portsToAssign.next(), port));
```
[the remark this answers is not in the document]
Comment (EronWright) [reply flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java]: A call to `takeRanges` won't give more than requested so it won't happen.

File: flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala [hunk 4]
```diff
@@ -185,7 +189,7 @@ class LaunchCoordinator(
         // process the assignments into a set of operations (reserve and/or launch)
         val slaveId = assignments.getLeasesUsed.get(0).getOffer.getSlaveId
         val offerIds = assignments.getLeasesUsed.asScala.map(_.getOffer.getId)
-        val operations = processAssignments(slaveId, assignments, remaining.toMap)
+        val operations = processAssignments(LOG, slaveId, assignments, remaining.toMap)
```
[the remark this answers is not in the document]
Comment (EronWright) [reply flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala]: The issue is that the class extends `Actor` which has a `log` field that we don't intend to use.  I will use `log` in `processAssignments`.

**Comment by EronWright** [comment]
@tillrohrmann I believe your feedback has been addressed, please take a look thanks.

File: flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/LaunchableMesosWorker.java]: Alright, thanks for the clarification.

File: flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala]: If I'm not mistaken, then the Scala `Buffer` implements `GenTraversableOnce` which is co-variant in its generic type and thus, there is no need for downcasting. I'll try it out and if it's possible, then I'll apply the changes while merging.

File: flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/LaunchCoordinator.scala]: alright.

**Comment by tillrohrmann** [comment]
Thanks for addressing this issue so quick @EronWright. Travis passed. Merging this PR.

**Comment by EronWright** [comment]
@tillrohrmann should we patch the 1.3 release too?   Or is 1.3.2 the last release there?

**Comment by tillrohrmann** [comment]
It depends a little bit on 1.4. We have serious build stability issues in the `release-1.3` branch and it would take a serious effort to stabilize it again. If I could choose, then we wouldn't have to release 1.3.3 because users switch directly to 1.4.

**Pull request closed without merging**